plaso/parsers/android_logcat.py

```diff
@@ -0,0 +1,225 @@
+      pyparsing.Or([
+          _ANDROID_FULL_DATE_GROUP,
+          _ANDROID_DATE_GROUP]).setResultsName('date') +
+      _ANDROID_TIME_GROUP.setResultsName('time') +
+      pyparsing.Optional(_ANDROID_TIME_ZONE).setResultsName('timezone') +
+      pyparsing.Word('VDIWEFS', exact=1).setResultsName('priority') +
+      pyparsing.Suppress('/') +
+      pyparsing.Word(
+          pyparsing.printables + ' ',
+          exclude_chars='(').setResultsName('tag') +
+      pyparsing.Suppress('(') +
+      pyparsing.Or([
+          pyparsing.Word(pyparsing.nums).setResultsName('pid'),
+          (pyparsing.Word(pyparsing.nums).setResultsName('uid') +
+           pyparsing.Suppress(':') +
+           pyparsing.Word(pyparsing.nums).setResultsName('pid'))]) +
+      pyparsing.Suppress(')') +
+      pyparsing.Suppress(': ') +
+      pyparsing.restOfLine.setResultsName('message'))
+
+  _BEGINNING_LINE = (
+      pyparsing.Suppress('--------- beginning of ') +
+      pyparsing.oneOf(['events', 'kernel', 'main', 'radio', 'system']))
+
+  LINE_STRUCTURES = [
+      ('beginning_line', _BEGINNING_LINE),
+      ('threadtime_line', _ANDROID_THREADTIME_LINE),
+      ('time_line', _ANDROID_TIME_LINE)]
+
+  _SUPPORTED_KEYS = frozenset([key for key, _ in LINE_STRUCTURES])
+
+  def ParseRecord(self, parser_mediator, key, structure):
+    """Parses a matching entry.
+
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+        and other components, such as storage and dfvfs.
+      key (str): name of the parsed structure.
+      structure (pyparsing.ParseResults): elements parsed from the file.
+
+    Raises:
+      ParseError: when the structure type is unknown.
+    """
+    if key not in self._SUPPORTED_KEYS:
+      raise errors.ParseError(
+          'Unable to parse record, unknown structure: {0:s}'.format(key))
+
+    if key == 'beginning_line':
+      return
+
+    if key in ('threadtime_line', 'time_line'):
+      parsed_time = self._GetValueFromStructure(structure, 'time')
+      parsed_date = self._GetValueFromStructure(structure, 'date')
+      parsed_timezone = self._GetValueFromStructure(structure, 'timezone')
+      if not parsed_timezone:
+        timezone = parser_mediator.timezone
+        parsed_timezone = datetime.datetime.now(timezone).strftime('%z')
+
+      parsed_timezone = parsed_timezone[:3] + ':' + parsed_timezone[3:]
+
+      if len(parsed_date) == 10:
+        date_format = f'{parsed_date}T{parsed_time}{parsed_timezone}'
+      else:
+        estimated_year = parser_mediator.GetEstimatedYear()
+        date_format = (f'{estimated_year}-{parsed_date}T'
+                       f'{parsed_time}{parsed_timezone}')
+
+      new_date_time = dfdatetime_time_elements.TimeElementsInMicroseconds()
+
+      try:
+        new_date_time.CopyFromStringISO8601(date_format)
+      except ValueError as error:
+        parser_mediator.ProduceExtractionWarning(
+            'Invalid date time value: {0:s}'.format(error))
```

Comment:
error => exception, an exception is not necessary an error
also use `{0!s}` since exception is not a string type